File: Lab1/module.cpp
#include <iostream>
#include <fstream>
#include <string>

using namespace std;

void createFile(const char* name) {
	ofstream f(name);
	string line;

	// errors check
	if (!f) {
		cerr << "Error: file \"" << name <<  "\" could not be opened" << endl;
		exit(1);
	}

	// read until any control character but space & EOL
	cout << "Enter your text, type any control character when finished" << endl;
	char c;
	while (cin.get(c) && (c > 31 || c == 10)) {
		f.put(c);
	}

	f.close();
}

char findFirstCharLastWord(string s) {
	s.erase(s.find_last_not_of(' ') + 1); //strip last spaces
	return s[s.find_last_of(' ') + 1];
}

void createNewFile(const char* name1, const char* name2) {
	ifstream inF(name1);
	ofstream outF(name2);
	string line;

	// errors check
	if (!inF || !outF) {
		cerr << "Error: file \"" << name1 << "\" or \"" << name2 << "\" could not be opened" << endl;
		exit(1);
	}

	getline(inF, line);
	while (1) {
		outF << findFirstCharLastWord(line) << line;
		if (getline(inF, line))  // if it is not the last line, add EOL, else end
			outF << endl;
		else
			break;
	}

	inF.close();
	outF.close();
}

void displayFile(const char* name) {
	string line;
	ifstream f(name);

	// errors check
	if (!f) {
		cerr << "Error: file \"" << name << "\" could not be opened" << endl;
		exit(1);
	}

	printf("\n----------\"%s\"----------\n", name);
	while (!f.eof()) {
		getline(f, line);
		cout << line << endl;
	}
	printf("----------%s----------\n\n", string(strlen(name) + 2, '-').c_str());

	f.close();
}	

File: lab_1/module.py
import sys

def get_user_input() -> str:
    """Turn the user's input into the text

    The input ends when a control character is entered
    """

    text = ""
    print("Enter your text, type any control character when finished")
    try:    #some of the control characters return error, handle it, the line is not gonna be saved
        while True:
            line = input()
            for c in line:
                if ord(c) > 31:    #if not a control char
                    text += c
                else:
                    return text    #exit
            text += '\n'
    except BaseException:
        return text
    

def create_file(name) -> None:
    """Create a file from the user input
    The input ends when a control character is entered

    Parameters
    ----------
    name : str
        The path to the file to be created
    """

    try:    # handle file opening errors
        with open(name, 'w') as f:
            f.write(get_user_input())
    except OSError:
        print(f"Error: file \"{name}\" could not be opened")
        sys.exit(1)

def create_new_file(name1, name2) -> None:
    """Create a new file getting the lines of the orig file and putting the first character of the last word in them in the beginning

    Parameters
    ----------
    name1 : str
        The path to the original file
    name2 : str
        The path to the file to be created
    """

    # handle file opening errors
    try:
        with open(name1, 'r') as inf:
            with open(name2, 'w') as outf:
                for line in inf.readlines():
                    if line.strip():    # check if not a blank line
                        line = line.strip().split()[-1][0] + line    #the first character of the last word in a line
                    outf.write(line)
    except OSError:
        print(f"Error: file \"{name1}\" or \"{name2}\" could not be opened")
        sys.exit(1)

def display_file(name) -> None:
    """Display the text in the file

    Parameters
    ----------
    name : str
        The path to the file
    """

    try:
        with open(name, 'r') as f:
            print('\n' + '-'*10 + name + '-'*10)
            print(f.read())
            print('-'*(20  + len(name)) + '\n')
    except OSError:
        print(f"Error: file \"{name}\" could not be opened")
        sys.exit(1)
